### ISSUE TITLE
Expensive deepcopy calls in SAM file parser

### ISSUE THREAD

**Issue by flaar94** [biopython/biopython#4825]
While trying to get a reasonably fast BAM file parser running working, I noticed that the sam file parser is a bit slow. With a little bit of investigation, it looks like over half of the time is being spent doing a deep copy of a SeqRecord on this line:
https://github.com/biopython/biopython/blob/22ebf951e736bd44f04e3552a4a39723ccb3c219/Bio/Align/sam.py#L702
![sam_reading_deepcopy](https://github.com/user-attachments/assets/15b57f8d-b97b-4985-ad88-bd96f1b2f499)

I dug into how it was created. Some of the `target` variables are actually just strings separate from the `target` Seq objects, but I think `self.targets[index]` only contains non-empty mutable attributes: `_seq: Seq` and `annotations: dict[str, str]`. `letter_annotations` is only added to the query object which is created from scratch, while `dbxrefs` and `features` don't appear in the file. 

So it seems like a depth 2 copy like:
```
target = copy.copy(self.targets[index])
target.seq = copy.copy(target.seq)
target.dbxrefs = target.dbxrefs[:]
target.features = target.features[:]
target.annotations = target.annotations.copy()
target.letter_annotations = target.letter_annotations.copy()
```
or building a new object like:
```
old_target = self.targets[index]
target = SeqRecord(copy.copy(old_target.seq), old_target.id, old_target.name, old_target.description, annotations=old_target.annotations.copy())
```
would give the same results.

Though even a tailored depth 3 copy/deepcopy like:
```
target = copy.copy(self.targets[index])
target.seq = copy.copy(target.seq)
target.dbxrefs = target.dbxrefs[:]
target.features = [copy.copy(feature) for feature in target.features]
target.annotations = {key: copy.copy(val) for key, val in target.annotations.items()}
target.letter_annotations = {key: copy.copy(val) for key, val in target.letter_annotations.items()}
```
is much more efficient than the built-in recursive deepcopy. 

The following is how long it took to read 200k alignments from a file for each modification
1. Original (copy.deepcopy): 22.3s
2. Depth 3 copy: 13.8s
3. Depth 2 copy: 13.1s
4. New Object: 11.7s

More generally, it might be worthwhile to add a built-in faster copying method to SeqRecord like:
```
def fast_copy(self):
    return self.__class__(copy.copy(self._seq), self.id, self.name, self.description,
                                self.dbxrefs[:],
                                [copy.copy(feature) for feature in self.features],
                                {key: copy.copy(val) for key, val in self.annotations.items()},
                                {key: copy.copy(val) for key, val in self._per_letter_annotations.items()})
```
or even dodge the redundant (and fairly slow) constructor validation:
```
@classmethod
def _from_validated(cls, seq, id, name, description, dbxrefs, features, annotations, letter_annotations):
    """constructor for post-validated data like copies"""
    inst = cls.__new__(cls)

    inst._seq = seq
    inst.id = id
    inst.name = name
    inst.description = description
    inst.dbxrefs = dbxrefs
    inst.features = features
    inst.annotations = annotations
    inst._per_letter_annotations = _RestrictedDict(length=len(seq))
    inst.letter_annotations = letter_annotations
    return inst

def fast_copy(self):
    return self._from_validated(copy.copy(self._seq), self.id, self.name, self.description,
                                self.dbxrefs[:],
                                [copy.copy(feature) for feature in self.features],
                                {key: copy.copy(val) for key, val in self.annotations.items()},
                                {key: copy.copy(val) for key, val in self._per_letter_annotations.items()})
```

### Setup

I am reporting a problem with Biopython version, Python version, and operating
system as follows:

```python
import sys; print(sys.version)
import platform; print(platform.python_implementation()); print(platform.platform())
import Bio; print(Bio.__version__)
```
Out:
```
5.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:27:10) [MSC v.1938 64 bit (AMD64)]
CPython Windows-10-10.0.22631-SP0
1.85.dev0
```

### Steps to reproduce

Here's the script I used for profiling. The data are the Sex Chromosomes of a IGSR sample (and maybe unmapped reads given the size of the file).

```
from Bio.Align.sam import AlignmentIterator
import time

path_to_sam = "D:\genomics_data\HG00138\sex_chrom\HG0013.sam"

start_time = time.perf_counter()
align_iter = AlignmentIterator(path_to_sam)
for i, line in enumerate(align_iter, 1):
    if i % 200_000 == 0:
        break
print(time.perf_counter() - start_time)
```

**Comment by michaelfm1211** [comment]
I think this is a good idea. For what's its worth here are my comments:

> More generally, it might be worthwhile to add a built-in faster copying method to SeqRecord like:
> ```py
> def fast_copy(self):
>     return self.__class__(copy.copy(self._seq), self.id, self.name, self.description,
>                                 self.dbxrefs[:],
>                                 [copy.copy(feature) for feature in self.features],
>                                 {key: copy.copy(val) for key, val in self.annotations.items()},
>                                 {key: copy.copy(val) for key, val in self._per_letter_annotations.items()})
> ```

From a quick glance over the `SeqRecord` class, it look like it doesn't currently implement a `__copy__()` method. Could we just use this as `__copy__()` then change the SAM code to use `target = copy.copy(self.targets[index])` (a shallow copy)? Would this work, or am I mistaken? IMO this looks a bit cleaner and idiomatic.

> or even dodge the redundant (and fairly slow) constructor validation:

Is there really any solid use case for `_from_validated()` outside of making copies? If not, then maybe this is should be part of the `fast_copy`/`__copy__()` code instead of creating a new method. Overall, the new `__copy__()` could look something like this:
```py
def __copy__(self):
    cls = self.__class__
    inst = cls.__new__(cls)

    inst._seq = copy.copy(self._seq)
    inst.id = self.id
    inst.name = self.name
    inst.description = self.description
    inst.dbxrefs = self.dbxrefs[:]
    inst.features = [copy.copy(features) for feature in self.features]
    inst.annotations = {key: copy.copy(val) for key, val in self.annotations.items()}
    inst._per_letter_annotations = _RestrictedDict(length=len(self.seq))
    inst.letter_annotations = {key: copy.copy(val) for key, val in self._per_letter_annotations.items()})
```

**Comment by flaar94** [comment]
**Overwriting built-in copy**
I think we could overwrite a built-in copy. There is a little bit of overhead from using magic methods, but nothing too huge (but this is a weird setting where most of the computational costs seem to be overhead). The only awkward thing is that the built-in copy magic methods are just for `copy.copy` which is supposed to be for shallow copies, and `copy.deepcopy` which is supposed to be for deep copies. This one is kind of in the middle. 

I think it's a bit closer to a deep copy, but someone could do something sketchy and add a letter annotation that is a sequence of lists, then get hard-to-track down bugs when they try to do a deep copy. It might be possible to do some checks to throw an error, but I'd have to see how much overhead that would add.


**In terms of the `_from_validated()` method:**
It's probably not necessary for this case, but it could be used in many of the built-in methods. `__add__`, `__radd__`, `reverse_complement`, `upper`,  slicing etc... all seem to do most of the the checking already (or it already being necessarily true because lengths don't change).

The current initialization produces a fair amount of overhead. For example, a big part of why `FastaIterator` is much slower than `SimpleFastaParser` is because initializing each SeqRecord is slow (it accounts for up about 40% of the time)

**Comment by mdehoon** [comment]
> The current initialization produces a fair amount of overhead. For example, a big part of why FastaIterator is much slower than SimpleFastaParser is because initializing each SeqRecord is slow (it accounts for about 40% of the time when I profiled it)

One solution is to implement the expensive attributes`dbxref`, `annotations`, `letter_annotations`, `features` of a `SeqRecord` as properties instead of attributes. Then they will get initialized only if they are actually used.

**Comment by flaar94** [comment]
I think something like that is probably a good idea for letter_annotations and `_RestrictedDict`. The current rough breakdown in initialization costs for initializing a `SeqRecord` with trivial mutable attributes is:

1. ~15% building a class (ie: `SeqRecord.__new__(SeqRecord`))
2. ~5% Function frame/magic of calling `SeqRecord()`/`__init__`
3. ~15% Validation if/then statements (particularly the `isinstance` calls)
4. ~50% initializing `_RestrictedDict`
5. ~15% Other Variable Initialization/Binding

I was looking into this earlier. Part of it is that initializing built-in objects is just way faster than building custom class instances, and another big chunk of time is taken by calling the parent class's init function. You can reduce that cost by about 1/3 by not inheriting and just wrapping the original object (also probably want to use slots to significantly reduce the memory overhead).

Strangely, the thing I tried that seemed to work the best while maintaining checks was to convert `_per_letter_annotations` as a built-in dict, and redefine the public facing getter for `letter_annotations` to build a thin wrapper instance which applies checks for `__setitem__` (kind of like pandas DataFrame `.loc` methods):
```
@property
def letter_annotations(self):
    return _RestrictedDict(len(self._seq), self._per_letter_annotations)
```

The full property setter (ie replacing all of  `_per_letter_annotations` instead of just an entry) could also be set to do the checks faster without this overhead or loops
```
@letter_annotations.setter
def letter_annotations(self, value):
    if not isinstance(value, dict):
        raise TypeError(
            "The per-letter-annotations should be a (restricted) dictionary."
        )
    if any(len(val) != len(self.seq) for val in value.values()):
        raise TypeError(
            "Any per-letter annotation should be a Python sequence "
            "(list, tuple or string) of the same length as the "
            f"biological sequence, here {len(self.seq)}."
        )  # This is better than the composition approach because the seq can change

    self._per_letter_annotations = value
```

```
class _RestrictedDict:
    _length: int
    _container: dict[str, Sequence[Any]]

    __slots__ = ("_length", "_container")

    def __init__(self, length, container):
        self._length = length
        self._container = container

    def __setitem__(self, key, value):
        if self._length != len(value):
            raise TypeError(
                "Any per-letter annotation should be a Python sequence "
                "(list, tuple or string) of the same length as the "
                f"biological sequence, here {self._length}."
            )

        self._container[key] = value

    def __getitem__(self, key):
        return self._container[key]

    def __getattr__(self, item):
        """This will catch any attributes that can't be inferred by other properties"""
        return getattr(self._container, item)
    
    ...
```
Unfortunately this would be a more significant change and might not be worth it (though it shouldn't affect the public API much), but I was curious how I could make it faster. It seemed to reduce init costs by 50%ish while not impacting dictionary entry assignment much.

Here were the specific results for 2 million iterations:
Empty Init
Time Elapsed: 1.885s -> Time Elapsed: 1.118s
Moderate Sized Init
Time Elapsed: 3.218s -> Time Elapsed: 2.313s
Letter Annotation Entry Assignment
Time Elapsed: 0.732s -> Time Elapsed: 0.755s
Full Letter Annotation Assignment
Time Elapsed: 2.002s -> Time Elapsed: 1.082s

And for FastaIterator:
14s -> 12s

Though if you bypass the iterator altogether with a custom method you get:
14s -> 9s

Anyways, though I think it's very interesting to learn what parts of Python have the most overhead, probably the main lesson I've taken away is that if we want to do specific things quickly, we're going to want to avoid big sequences of Python classes. Maybe an eventual solution is to have wrappers for the entire sequence, maybe like a SeqRecordArray, and internally it could be something like a numpy record array, though public facing it could act like a list of SeqRecords or similar

**Comment by flaar94** [comment]
Oh, I forgot to mention. I also looked at setting up letter_annotations more directly as a cached property using various approaches, more directly like your suggestion. Unfortunately, if I remember correctly, it produced a lot more overhead on the getter unfortunately

**Comment by peterjc** [comment]
I was always a bit frustrated with the object creation overheads with using FASTQ via SeqRecord objects - you're exploring techniques I never considered at the time, although some of the parsers do avoid the length checking (since the parser has already done it) by using the private attributes directly.

**Comment by mdehoon** [comment]
@flaar94  Are you planning to make a PR for this?

**Comment by flaar94** [comment]
@mdehoon yeah, see the above draft PR.  I was fighting with pre-commit. My black seems to be set up differently from pre-commit's settings. I'll figure it out later. 

Anyways, this version is relatively pared down, mostly just adding a copy method to SeqRecord and SeqFeature (the latter because it speeds up and simplifies the copy method in SeqRecord). I included the _from_validated alternative constructor, which at least avoids checks on the letter_annotations.

To get improvements on initializing sequences without letter_annotations I think we'll need to change _RestrictedDict. Given the headaches I discovered involved with trying to avoid the slow subclass system, I think it might be worth switching to composition like the above version I posted, even if we don't change the getter. It's quite a bit faster, and uses less memory. I can dig into the attribute dictionary and make sure all the public methods look identical after the change

**Comment by flaar94** [comment]
I think mdehoon's idea was probably best after all. I was able to change letter_annotations to only get initialized after the setter or getter gets called without any clear performance regressions. Since I was changing them, I also changed their property formatting to fit modern property decorator style. Unfortunately, this attracted mypy's type-checking gaze, so I spent a good amount of time appeasing it.

The biggest complaint seemed to be that seq_record.seq can be None, but most of the methods assume it's a Seq or MutableSeq. What is the intended usecase there? Is it supposed to be an empty sequence or an unknown seq or something else? I'm not sure if the I handled that case is correct

(Other thorny type issues seemed to involve mypy not liking that the setter and getter for a property had different types, but I found the issue thread and they seem to be in the process of fixing that. I just silenced that complaint, along with a couple that should be caught using try: except: blocks, or mypy not recognizing that dict is a MutableMapping)

**Comment by mdehoon** [comment]
> The biggest complaint seemed to be that seq_record.seq can be None, but most of the methods assume it's a Seq or MutableSeq. What is the intended use case there? 

To find out, you could try disallowing record.seq = None, and see which of the tests fail.

**Comment by flaar94** [comment]
Good idea. I also did some searches for `SeqRecord(None`. It seems to primarily be used internally parsers as a way to incrementally build up SeqRecords

**Comment by flaar94** [comment]
Ok, given that, I'm going to err on the side of keeping the current behavior and just throwing errors, because it probably represents someone letting a SeqRecord slip out without finishing it. I'll just catch the error messages to make it more clear what went wrong. We can always extend those methods to the None case later if we need it for some application
